cache/remotecache/v1/parse.go

```diff
@@ -64,6 +64,33 @@ func parseRecord(cc CacheConfig, idx int, provider DescriptorProvider, t solver.
 			r.AddResult(res.CreatedAt, remote)
 		}
 	}
+	// TODO: what the behavior on previous builders when there's no Result. Is it
+	// assumed that this record is Scratch? Or is it a cache miss? Want the latter.
```

Comment:
@tonistiigi This is my main remaining question, the answer wasn't immediately obvious when jumping through the code so would appreciate any insight you can provide.

This case is important because now when a record can't be specified using the old style `CacheResult` and instead has to use the new `DiffResults`, and old buildkitd will see that the record exists but has empty `CacheResult`. The behavior I want is for these buildkitds to then just get a cache miss in this case.

If it turns out instead that they assume "no result" means "result is scratch" then we'll have to modify this slightly. We could add a new type to replace `CacheRecord` that exclusively uses `DiffResults` (so old builders won't even see records that use diff results). Or another option would be to have `CacheRecord` set its `LayerIndex` to `-1` in these cases, which would mean that old buildkitds get an error when trying to import this cache. That's not ideal obviously but is better than them getting a cache hit with the incorrect result.

Reply:
If you never call `AddResult` then it will not load any cache. Loading scratch should be possible with `AddResult(nil)` (maybe). In cache interface https://github.com/moby/buildkit/blob/master/docs/solver.md#cache-providers `query` and `records()` are separate concepts. So a vertex can match cache but there is no loadable record behind that match. This is because we can use the cache match to check for a cache match for the next vertex and that might be loadable now.